### PR TITLE
Add migration support for block devices

### PR DESCRIPTION
**Pull request by wallyworld**
As well as adding support to allow block devices to be exported/imported from/to models, this PR also fixes the DDL for the block device to machine association.

The association between block devices and their machine was being done using a many-many join table. This PR changes that to introduce a machine FK into the block device table to model as a 1 to many. It also adds more filesystem types to the filesystem type look up table.

The modelmigration package adds support for migrating block devices. This required an updated description package to allow block devices to be added to the model separately to the machines which host them. The service package gains the necessary methods and unit tests to support all of this, plus the upcoming replacement of block devices in mongo.

## QA steps

For now, just the unit tests.
The next PR will wire it all up.
Also needs https://github.com/juju/juju/pull/16797 to land for end to end testing.

## Links

**Jira card:** JUJU-5335

### CODE REVIEW

File: domain/blockdevice/modelmigration/export.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+
+	"github.com/juju/juju/core/modelmigration"
+	"github.com/juju/juju/domain/blockdevice"
+	"github.com/juju/juju/domain/blockdevice/service"
+	"github.com/juju/juju/domain/blockdevice/state"
+)
+
+var logger = loggo.GetLogger("juju.migration.blockdevices")
```
Comment (SimonRichardson) [inline domain/blockdevice/modelmigration/export.go:19]: I'll sweep back through the migration code and add a logger.

File: domain/blockdevice/modelmigration/export.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+type exportOperation struct {
+	modelmigration.BaseOperation
+
+	service ExportService
+}
+
+// Setup implements Operation.
+func (e *exportOperation) Setup(scope modelmigration.Scope) error {
+	// We must not use a watcher during migration, so it's safe to pass a
+	// nil watcher factory.
```
Comment (SimonRichardson) [inline domain/blockdevice/modelmigration/export.go:43]: Thanks for the reminder I'll fix this when I do the logger.

File: domain/blockdevice/modelmigration/export_test.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	c.Assert(bd.Label(), gc.Equals, "label")
+	c.Assert(bd.UUID(), gc.Equals, "device-uuid")
+	c.Assert(bd.HardwareID(), gc.Equals, "hardware-id")
+	c.Assert(bd.WWN(), gc.Equals, "wwn")
+	c.Assert(bd.BusAddress(), gc.Equals, "bus-address")
+	c.Assert(bd.SerialID(), gc.Equals, "serial-id")
+	c.Assert(bd.Size(), gc.Equals, uint64(100))
+	c.Assert(bd.FilesystemType(), gc.Equals, "ext4")
+	c.Assert(bd.InUse(), jc.IsTrue)
+	c.Assert(bd.MountPoint(), gc.Equals, "/path/to/here")
```
Comment (SimonRichardson) [inline domain/blockdevice/modelmigration/export_test.go:90]: nit: These should be `c.Check`

File: domain/blockdevice/service/service.go [hunk 4]
```diff
@@ -56,6 +60,34 @@ func NewService(st State, wf WatcherFactory, logger Logger) *Service {
...
+// BlockDevices returns the block devices for a specified machine.
+func (s *Service) BlockDevices(ctx context.Context, machineId string) ([]blockdevice.BlockDevice, error) {
+	return s.st.BlockDevices(ctx, machineId)
+}
+
+// UpdateBlockDevices updates the block devices for a specified machine.
+func (s *Service) UpdateBlockDevices(ctx context.Context, machineId string, devices ...blockdevice.BlockDevice) error {
+	for i := range devices {
+		if devices[i].FilesystemType == "" {
+			devices[i].FilesystemType = "unspecified"
```
Comment (SimonRichardson) [inline domain/blockdevice/service/service.go:72]: Can we make `"unspecified"` a constant with a comment.

File: domain/blockdevice/state/state.go [hunk 5]
```diff
@@ -288,42 +272,64 @@ func blockDevicesChanged(oldDevices, newDevices []blockdevice.BlockDevice) bool
...
-		}
-		machineUUID := result["machine_uuid"].(string)
-		if err := RemoveMachineBlockDevices(ctx, tx, machineUUID); err != nil {
-			return errors.Annotatef(err, "removing block devices on machine %q (%s)", machineId, machineUUID)
+	types := []any{
+		BlockDevice{},
+		FilesystemType{},
+		DeviceLink{},
+		BlockDeviceMachine{},
+	}
```
Comment (SimonRichardson) [inline domain/blockdevice/state/state.go:298]: This is not nice at all. I wonder if it's more palatable to be outside of the method and just be a package variable.

File: domain/schema/model.go [hunk 6]
```diff
@@ -431,10 +435,15 @@ CREATE UNIQUE INDEX idx_filesystem_type_name
...
-    (3, 'zfs');
+    (0, 'unspecified'),
+    (1, 'vfat'),
+    (2, 'ext4'),
+    (3, 'xfs'),
+    (4, 'btrfs'),
+    (5, 'zfs'),
+    (6, 'jfs'),
+    (7, 'squashfs'),
+    (8, 'bcachefs');
```
Comment (SimonRichardson) [inline domain/schema/model.go:446]: Lovely.

File: domain/blockdevice/state/state.go (hunk 5 above)
Comment (wallyworld) [reply domain/blockdevice/state/state.go:298]: I don't see the issue. The sqlair Prepare() method takes a variadac list of types. It's convenient to assign them to to a var next to the call to Prepare() to keep the code a bit more organised. We do this is a few places.

File: domain/blockdevice/modelmigration/export_test.go (hunk 3 above)
Comment (wallyworld) [reply domain/blockdevice/modelmigration/export_test.go]: done

File: domain/blockdevice/service/service.go (hunk 4 above)
Comment (wallyworld) [reply domain/blockdevice/service/service.go]: done

**Pull request merged**